manuals/vsapminfo.1
.Dd June 26, 2026
.Dt VSAPMINFO 1
.Os
.Sh NAME
.Nm vsapminfo
.Nd determines information about an Apple Partition Map (APM) volume system
.Sh SYNOPSIS
.Nm vsapminfo
.Op Fl b Ar bytes_per_sector
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm vsapminfo
is a utility to determine information about an Apple Partition Map (APM) \
volume system
.Pp
.Nm vsapminfo
is part of the
.Nm libvsapm
package.
.Nm libvsapm
is a library to access the Apple Partition Map (APM) volume system format
.Pp
.Ar source
is the source image.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl b Ar bytes_per_sector
specify the number of bytes per sector (default is 512).
Use this to override the automatic bytes per sector detection
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# vsapminfo /dev/sda
vsapminfo 20260626
.sp
Apple Partition Map (APM) information:
	Bytes per sector	: 512
	Number of partitions	: 2
.sp
Partition: 1
	Type			: Apple_HFS
	Name			: Foo
	Offset			: 33792 (0x00008400)
	Size			: 228352
	Status flags		: 0x0000007f
		Is valid
		Is allocated
		Is in use
		Contains boot information
		Is readable
		Is writeable
		Boot code is position independent
.sp
Partition: 2
	Type			: Apple_Free
	Name			: Extra
	Offset			: 32768 (0x00008000)
	Size			: 1024
	Status flags		: 0x00000000
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libvsapm/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* pyvsapm: fix libvsapm_partition_get_type

* finalize a first version of the library and Python bindings
* detect leading drive descriptor

* determine what can generate a hybrid disk that contains both an ISO 9660 primary volume descriptor and an Apple Partition Map

20210509
* see `git log' for more recent change log
* initial version based on previous work

20090201
* created documentation



vsapmtools/vsapminfo.c
/*
 * Shows information obtained from an Apple Partition Map (APM) volume system.
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "vsapmtools_getopt.h"
#include "vsapmtools_libcerror.h"
#include "vsapmtools_libclocale.h"
#include "vsapmtools_libcnotify.h"
#include "vsapmtools_libvsapm.h"
#include "vsapmtools_output.h"
#include "vsapmtools_signal.h"
#include "vsapmtools_unused.h"

info_handle_t *vsapminfo_info_handle = NULL;
int vsapminfo_abort                  = 0;

/* Signal handler for vsapminfo
 */
void vsapminfo_signal_handler(
      vsapmtools_signal_t signal VSAPMTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "vsapminfo_signal_handler";

	VSAPMTOOLS_UNREFERENCED_PARAMETER( signal )

	vsapminfo_abort = 1;

	if( vsapminfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     vsapminfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use vsapminfo to determine information about an Apple Partition Map (APM) volume system.";

	vsapmtools_option_t options[ ] = {
		{ 'b', "bytes_per_sector", "specify the number of bytes per sector (default is 512). Use this to override the automatic bytes per sector detection" },
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source image" },
	};
	system_character_t options_string[ 32 ];

	libvsapm_error_t *error                     = NULL;
	system_character_t *option_bytes_per_sector = NULL;
	system_character_t *source                  = NULL;
	char *program                               = "vsapminfo";
	system_integer_t option                     = 0;
	int number_of_options                       = (int) ( sizeof( options ) / sizeof( vsapmtools_option_t ) );
	int result                                  = 0;
	int verbose                                 = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "vsapmtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( vsapmtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	vsapmtools_output_version_fprint(
	 stdout,
	 program );

	if( vsapmtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = vsapmtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				vsapmtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'b':
				option_bytes_per_sector = optarg;

				break;

			case (system_integer_t) 'h':
				vsapmtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				vsapmtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source image.\n" );

		vsapmtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libvsapm_notify_set_stream(
	 stderr,
	 NULL );
	libvsapm_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &vsapminfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( vsapminfo_info_handle != NULL );
#endif
	if( option_bytes_per_sector != NULL )
	{
		result = info_handle_set_bytes_per_sector(
			  vsapminfo_info_handle,
			  option_bytes_per_sector,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set bytes per sector.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported bytes per sector defaulting to: %" PRIu32 ".\n",
			 vsapminfo_info_handle->bytes_per_sector );
		}
	}
	if( info_handle_open_input(
	     vsapminfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source image.\n" );

		goto on_error;
	}
	if( info_handle_partitions_fprint(
	     vsapminfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print partitions information.\n" );

		goto on_error;
	}
	if( info_handle_close_input(
	     vsapminfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &vsapminfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( vsapminfo_info_handle != NULL )
	{
		info_handle_free(
		 &vsapminfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

